src/main/scala/carts/cart/Repository.scala

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright © 2022-2023 by Pervasive Cats S.r.l.s.
+ *
+ * All Rights Reserved.
+ */
+
+package io.github.pervasivecats
+package carts.cart
+
+import scala.util.Try
+
+import com.typesafe.config.Config
+import eu.timepit.refined.auto.autoUnwrap
+import eu.timepit.refined.auto.given
+import io.getquill.*
+
+import carts.cart.valueobjects.{CartId, Customer, Store}
+import carts.cart.valueobjects.Customer.WrongCustomerFormat
+import carts.{Validated, ValidationError}
+import carts.cart.entities.{AssociatedCart, Cart, LockedCart, UnlockedCart}
+import AnyOps.*
+
+trait Repository {
+
+  def findById(cartId: CartId): Validated[Cart]
+
+  def findByStore(store: Store): Validated[Set[Cart]]
+
+  def add(store: Store): Validated[LockedCart]
+
+  def update(cart: Cart): Validated[Unit]
+
+  def remove(cart: Cart): Validated[Unit]
+}
+
+object Repository {
+
+  case object OperationFailed extends ValidationError {
+
+    override val message: String = "The operation on the given cart has failed"
+  }
+
+  case object CartAlreadyPresent extends ValidationError {
+
+    override val message: String = "The cart is already present"
+  }
+
+  case object CartNotFound extends ValidationError {
+
+    override val message: String = "The queried cart was not found"
+  }
+
+  private class PostgresRepository(ctx: PostgresJdbcContext[SnakeCase]) extends Repository {
+
+    import ctx.*
+
+    private case class Carts(cartId: Long, store: Long, movable: Boolean, customer: Option[String])
+
+    private def protectFromException[A](f: => Validated[A]): Validated[A] = {
+      Try(f).getOrElse(Left[ValidationError, A](OperationFailed))
+    }
+
+    private def queryById(cartId: CartId) = quote {
+      query[Carts].filter(c => c.cartId === lift[Long](cartId.value))
+    }
+
+    private def validateCart(cart: Carts): Validated[Cart] =
+      cart
+        .customer
+        .fold(
+          for {
+            cartId <- CartId(cart.cartId)
+            store <- Store(cart.store)
+          } yield (cartId, store, cart.movable) match {
```

Comment:
This can be refactored to for { ... } yield if (cart.isMovable) UnlockedCart(...) else LockedCart(...)